File: src/main/java/br/com/gabrielalmir/gotcha/utils/Snowflake.java
package br.com.gabrielalmir.gotcha.utils;

public class Snowflake {
    private static final long EPOCH = 1288834974657L;
    private static final long WORKER_ID_BITS = 5L;
    private static final long DATACENTER_ID_BITS = 5L;
    private static final long SEQUENCE_BITS = 12L;

    private static final long MAX_WORKER_ID = ~(-1L << WORKER_ID_BITS);
    private static final long MAX_DATACENTER_ID = ~(-1L << DATACENTER_ID_BITS);
    private static final long SEQUENCE_MASK = ~(-1L << SEQUENCE_BITS);

    private static final long WORKER_ID_SHIFT = SEQUENCE_BITS;
    private static final long DATACENTER_ID_SHIFT = SEQUENCE_BITS + WORKER_ID_BITS;
    private static final long TIMESTAMP_LEFT_SHIFT = SEQUENCE_BITS + WORKER_ID_BITS + DATACENTER_ID_BITS;

    private long workerId;
    private long datacenterId;
    private long sequence = 0L;
    private long lastTimestamp = -1L;

    public Snowflake(long workerId, long datacenterId) {
        if (workerId > MAX_WORKER_ID || workerId < 0) {
            throw new IllegalArgumentException(String.format("Worker ID deve estar entre 0 e %d", MAX_WORKER_ID));
        }
        if (datacenterId > MAX_DATACENTER_ID || datacenterId < 0) {
            throw new IllegalArgumentException(String.format("Datacenter ID deve estar entre 0 e %d", MAX_DATACENTER_ID));
        }
        this.workerId = workerId;
        this.datacenterId = datacenterId;
    }

    public synchronized long nextId() {
        long timestamp = currentTime();

        if (timestamp < lastTimestamp) {
            throw new RuntimeException("Clock está indo para trás. Recusando gerar ID.");
        }

        if (lastTimestamp == timestamp) {
            sequence = (sequence + 1) & SEQUENCE_MASK;

            if (sequence == 0) {
                timestamp = waitNextMillis(lastTimestamp);
            }
        } else {
            sequence = 0;
        }

        lastTimestamp = timestamp;

        return ((timestamp - EPOCH) << TIMESTAMP_LEFT_SHIFT)
                | (datacenterId << DATACENTER_ID_SHIFT)
                | (workerId << WORKER_ID_SHIFT)
                | sequence;
    }

    private long waitNextMillis(long lastTimestamp) {
        long timestamp = currentTime();
        while (timestamp <= lastTimestamp) {
            timestamp = currentTime();
        }
        return timestamp;
    }

    private long currentTime() {
        return System.currentTimeMillis();
    }
}

File: src/utils/snowflake.ts
export class Snowflake {
    private static readonly EPOCH = 1288834974657;
    private static readonly WORKER_ID_BITS = 5;
    private static readonly DATACENTER_ID_BITS = 5;
    private static readonly SEQUENCE_BITS = 12;

    private static readonly MAX_WORKER_ID = ~(-1 << Snowflake.WORKER_ID_BITS);
    private static readonly MAX_DATACENTER_ID = ~(-1 << Snowflake.DATACENTER_ID_BITS);
    private static readonly SEQUENCE_MASK = ~(-1 << Snowflake.SEQUENCE_BITS);

    private static readonly WORKER_ID_SHIFT = Snowflake.SEQUENCE_BITS;
    private static readonly DATACENTER_ID_SHIFT = Snowflake.SEQUENCE_BITS + Snowflake.WORKER_ID_BITS;
    private static readonly TIMESTAMP_LEFT_SHIFT = Snowflake.SEQUENCE_BITS + Snowflake.WORKER_ID_BITS + Snowflake.DATACENTER_ID_BITS;

    private workerId: number;
    private datacenterId: number;
    private sequence: number = 0;
    private lastTimestamp: number = -1;

    constructor(workerId: number, datacenterId: number) {
        if (workerId > Snowflake.MAX_WORKER_ID || workerId < 0) {
            throw new Error(`Worker ID deve estar entre 0 e ${Snowflake.MAX_WORKER_ID}`);
        }
        if (datacenterId > Snowflake.MAX_DATACENTER_ID || datacenterId < 0) {
            throw new Error(`Datacenter ID deve estar entre 0 e ${Snowflake.MAX_DATACENTER_ID}`);
        }
        this.workerId = workerId;
        this.datacenterId = datacenterId;
    }

    public nextId(): number {
        let timestamp = this.currentTime();

        if (timestamp < this.lastTimestamp) {
            throw new Error("Clock está indo para trás. Recusando gerar ID.");
        }

        if (this.lastTimestamp === timestamp) {
            this.sequence = (this.sequence + 1) & Snowflake.SEQUENCE_MASK;

            if (this.sequence === 0) {
                timestamp = this.waitNextMillis(this.lastTimestamp);
            }
        } else {
            this.sequence = 0;
        }

        this.lastTimestamp = timestamp;

        return ((timestamp - Snowflake.EPOCH) << Snowflake.TIMESTAMP_LEFT_SHIFT)
            | (this.datacenterId << Snowflake.DATACENTER_ID_SHIFT)
            | (this.workerId << Snowflake.WORKER_ID_SHIFT)
            | this.sequence;
    }

    private waitNextMillis(lastTimestamp: number): number {
        let timestamp = this.currentTime();
        while (timestamp <= lastTimestamp) {
            timestamp = this.currentTime();
        }
        return timestamp;
    }

    private currentTime(): number {
        return Date.now();
    }
}
